draft-barnes-hpke-hpke.md

```diff
@@ -649,9 +674,26 @@ def VerifyPSKInputs(mode, psk, psk_id):
   if (not got_psk) and (mode in [mode_psk, mode_auth_psk]):
     raise Exception("Missing required PSK input")
 
-def KeySchedule<ROLE>(mode, shared_secret, info, psk, psk_id):
-  VerifyPSKInputs(mode, psk, psk_id)
+# For use with a single-stage KDF
+def CombineSecrets(mode, shared_secret, info, psk, psk_id):
+  derive_input = concat(
+    lengthPrefixed(psk),
+    lengthPrefixed(shared_secret),
+    mode,
+    lengthPrefixed(psk_id),
+    lengthPrefixed(info),
+  )
 
+  secret = LabeledDerive(derive_input, "secret", Nk + Nn + Nh)
```

Comment:
You are doing the single invocation and slicing of output to avoid overheads, is that right?

Did you look into doing:

```
key = LabeledDerive(derive_input, "key", Nk)
base_nonce = LabeledDerive(derive_input, "base_nonce", Nn)
exporter_secret = LabeledDerive(derive_input, "exp", Nh)
```
?

Reply:
Yes, the hpke-bench did look at structure like that. It is ~30% overhead for the main KDF, even with optimal caching.